runtime/Cpp/runtime/src/misc/Interval.h

```diff
@@ -13,27 +13,30 @@ namespace misc {
   // Helpers to convert certain unsigned symbols (e.g. Token::EOF) to their original numeric value (e.g. -1)
   // and vice versa. This is needed mostly for intervals to keep their original order and for toString()
   // methods to print the original numeric value (e.g. for tests).
-  size_t numericToSymbol(ssize_t v);
-  ssize_t symbolToNumeric(size_t v);
+  constexpr size_t numericToSymbol(ssize_t v) { return static_cast<size_t>(v); }
+
+  constexpr ssize_t symbolToNumeric(size_t v) { return static_cast<ssize_t>(v); }
 
   /// An immutable inclusive interval a..b
-  class ANTLR4CPP_PUBLIC Interval {
+  class ANTLR4CPP_PUBLIC Interval final {
   public:
     static const Interval INVALID;
 
     // Must stay signed to guarantee the correct sort order.
     ssize_t a;
     ssize_t b;
 
-    Interval();
-    explicit Interval(size_t a_, size_t b_); // For unsigned -> signed mappings.
-    Interval(ssize_t a_, ssize_t b_);
+    constexpr Interval() : Interval(static_cast<ssize_t>(-1), static_cast<ssize_t>(-2)) {}
+
+    constexpr explicit Interval(size_t a_, size_t b_) : Interval(symbolToNumeric(a_), symbolToNumeric(b_)) {}
+
+    constexpr Interval(ssize_t a_, ssize_t b_) : a(a_), b(b_) {}
 
     /// return number of elements between a and b inclusively. x..x is length 1.
     ///  if b < a, then length is 0.  9..10 has length 2.
-    size_t length() const;
+    constexpr size_t length() const { return b >= a ? static_cast<size_t>(b - a + 1) : 0; }
 
-    bool operator == (const Interval &other) const;
+    constexpr bool operator==(const Interval &other) const { return a == other.a && b == other.b; }
```

Comment:
I'm not so fond of having implementation code in header files. One single change (and be it for testing) and you might have to recompile everything that includes that header (in the worst case the entire application, which can take a while in C++ as you know).

Reply:
Putting the implementation in the header allows the compiler to perform optimizations without requiring link time optimization. Without it in the header file, the compiler is forced to emit a function call. With it in the header file, the compiler can boil down the Interval constructor to two simple move instructions.

I am not saying we should put everything in the header, but rather code that is called frequently and is relatively simple. Mostly accessors, logical operators, and etc. Things that boil down to a few instructions or simply forward to another function.

Reply:
OK, I follow your explanation.